src/lib.rs

```diff
@@ -1,9 +1,361 @@
+    /// to every connected peer for whom `Share::should_share` returns `ShareDecision::Share`
+    ///
+    /// # Errors
+    ///
+    /// * If a document with this ID already exists in storage
+    pub async fn create(&self, doc: am::Automerge) -> Result<DocHandle, error::Create<Store>> {
+        todo!()
+    }
+
+    /// Announce that we have a given document and wait for the peers specified by `ack_by` to sync
+    /// with us
+    ///
+    /// # Errors
+    ///
+    /// * If the given document does not exist
+    /// * If there is no connection to the peers given in `ack_by`
+    pub async fn announce(&self, doc: DocumentId, ack_by: PeerSpec) -> Result<(), error::Announce> {
+        todo!()
+    }
+
+    /// Shut down the repository
+    ///
+    /// This will signal to the background task that it should shutdown and
+    /// then wait for it to complete
+    ///
+    /// If the background task has already shutdown this returns immediately. If
+    /// another task has already signalled the background task to shutdown then
+    /// this will wait for shutdown to complete.
+    pub async fn shutdown(self) -> Result<(), error::Shutdown> {
+        todo!()
+    }
+
+    /// Get the currently connected peers
+    pub async fn peers(&self) -> Vec<PeerInfo> {
+        todo!()
+    }
+
+    /// A stream of events about the peers connected to this repo
+    pub fn peer_events(&self) -> impl Stream<Item = PeerEvent> {
+        futures::stream::empty()
+    }
+
+    /// Wait for the peers specified by `peers` to be connected
+    pub async fn peers_connected(&self, peers: PeerSpec) {}
+}
+
+/// The messages of the multi-document sync protocol
+///
+/// The multi-doc sync protocol works like this:
+///
+/// 1. The connecting peer sends a `Message::Join` containing its peer ID
+/// 2. The accepting peer sends a `Message::Peer` containing its peer ID
+/// 3. Any time a peer wishes to send a sync for a particular document it sends
+///    a `Message::Message` with a channel ID of `ChannelId::Document`
+/// 3. Any time a peer wishes to send ephemeral data for a particular document
+///    it sends a `Message::Message` with a channel ID of `ChannelId::Ephemeral`
+pub enum Message {
+    /// Sent by the connecting peer on opening a connection to tell the other
+    /// end their peer ID
+    Join {
+        /// Must always be `ChannelId::Sync`
+        channel_id: ChannelId,
+        /// The peer ID of the sender
+        sender_id: PeerId,
+    },
+    /// A sync message for a particular document
+    Message {
+        channel_id: ChannelId,
+        sender_id: PeerId,
+        target_id: PeerId,
+        /// The data of the sync message
+        data: Vec<u8>,
+    },
+    /// Sent by the accepting peer after having received [`Join`] to tell the
+    /// connecting peer their peer ID.
+    Peer {
+        /// The PeerId of the accepting peer
+        sender_id: PeerId,
+    },
+}
```

Comment:
I'd say these are good message to send between network peers, but between repos I would reduce the communication to the one in the middle(and with the actual deserialized sync message). 

This is really a discussion about what `connect` looks like. It seems to me it can either be:
1. Connect with a source(typically a socket address), then send a Join(Id) message to identify the remote repo, send a Peer(id) back, and then start syncing. 
2.  Connect with an PeerId, start syncing. 

If there is a need to do 1, I would rather do it in a component that is separate from the repo. A kind of networking front-end that would service raw `Source` based connection, try to complete the handshake within some timeout, and then for a given PeerId pass a stream of messages to the repo(and close the stream if the source disconnects).

Reply:
What's the advantage of moving this into a separate component vs the user just having a timeout on the whole `connect` future? The disadvantage is obviously the API becomes more complex, the user now has to think about handshakes.

Reply:
I mean we should build an utility that does this, which would handle the initial handshake and then pass a sink/stream to the repo once it succeeds. Then users could still use the raw underlying API to build their own if needed. 

The advantage is more separation of concern, also the ability to ringfence the async part. I can imagine a simple piece of code that runs a loop and handles incoming socket addrs, then below that we have the repo that runs the core logic.